db/version_builder.cc

```diff
@@ -517,6 +517,28 @@ class VersionBuilder::Rep {
     return meta->oldest_blob_file_number;
   }
 
+  uint64_t GetMinOldestBlobFileNumber() const {
+    uint64_t min_oldest_blob_file_num = std::numeric_limits<uint64_t>::max();
+    for (int level = 0; level < num_levels_; ++level) {
+      const auto& base_files = base_vstorage_->LevelFiles(level);
+      for (const auto* fmeta : base_files) {
+        assert(fmeta);
+        min_oldest_blob_file_num =
+            std::min(min_oldest_blob_file_num, fmeta->oldest_blob_file_number);
+      }
+      const auto& added_files = levels_[level].added_files;
+      for (const auto& elem : added_files) {
+        assert(elem.second);
+        min_oldest_blob_file_num = std::min(
+            min_oldest_blob_file_num, elem.second->oldest_blob_file_number);
+      }
```

Comment:
> Unrelated to the above but I think another approach would be to keep a running `min_oldest_blob_file_num` instead of recomputing upon each call. That is, we could maintain `min_oldest_blob_file_num` in a member variable, which could be updated as table files are added or removed.

iiuc, in order to maintain a running `min_oldest_blob_file_num`, we need to maintain the map from blob file number to a set of table files whose oldest blob reside in this blob file. This information is available in `VersionStorageInfo::blob_files_`. However, until we call `SaveTo()` when actually creating a version, this information is not readily available in `VersionBuilder`. Is this correct?

Reply:
> iiuc, in order to maintain a running `min_oldest_blob_file_num`, we need to maintain the map from blob file number to a set of table files whose oldest blob reside in this blob file. This information is available in `VersionStorageInfo::blob_files_`. However, until we call `SaveTo()` when actually creating a version, this information is not readily available in `VersionBuilder`. Is this correct?

Well, I guess that depends on the definition of "readily available" :) but it's available in `blob_file_meta_deltas_`. Thinking about this a little bit more, I believe we don't actually need to maintain a running `min_oldest_blob_file_num` in `VersionEditHandler`. Instead, we could make `GetMinOldestBlobFileNumber` a bit more efficient: we could return the smaller of the lowest blob file number from the base version and the lowest blob file number from `blob_file_meta_deltas_` which has linked SSTs. Both affected containers are `std::map`s so I think this could be done pretty efficiently. What do you think?

Reply:
Just saw you updated the comment. Yes, that's what I was thinking.

Reply:
> Before we update `min_oldest_blob_file_num`, shouldn't we also make sure that the table file hasn't been removed? During recovery, several `VersionEdits` are accumulated in the same `VersionEdit`, so a later `DeleteFile` can cancel out a `NewFile`.

Not sure if I get this.
Here, the purpose is to compute `min_oldest_blob_file_num` up to this point, not considering later version edits. The `added_files` contain table files that are not obsolete now. We call this method before applying the upcoming version edit. By calling this method and comparing the result with the tracked missing_blob_file_high for the cf, we know whether we have missing blob files before applying the version edit. Later, if we see the missing_blob_file_high has not decreased, we pessimistically assume that we will miss blob files.

Reply:
> > Before we update `min_oldest_blob_file_num`, shouldn't we also make sure that the table file hasn't been removed? During recovery, several `VersionEdits` are accumulated in the same `VersionEdit`, so a later `DeleteFile` can cancel out a `NewFile`.
> 
> Not sure if I get this.
> Here, the purpose is to compute `min_oldest_blob_file_num` up to this point, not considering later version edits. The `added_files` contain table files that are not obsolete now. We call this method before applying the upcoming version edit. By calling this method and comparing the result with the tracked missing_blob_file_high for the cf, we know whether we have missing blob files before applying the version edit. Later, if we see the missing_blob_file_high has not decreased, we pessimistically assume that we will miss blob files.

What I meant here is that after applying a series of `VersionEdit`s, it is possible that a file that was added a while ago has already been removed (by a compaction). In that case, it would appear both in `added_files` and `deleted_files` for the level. At that point, the table file is obsolete, and thus it is OK for the blob file it points to to be "missing," right? (In other words, obsolete table files should not participate in the calculation of `min_oldest_blob_file_num`.)

Reply:
I think the above would work for trivially moved files as well, since they would appear in the `added_files` list for the new level

Reply:
> after applying a series of VersionEdits, it is possible that a file that was added a while ago has already been removed

Right. According to the implementation of `VersionBuilder::Rep::ApplyFileDeletion()`, if a file is deleted, then it must not be in `added_files`. When we compute min_oldest_blob_file_number, we iterate over the files that **currently** exist in version.
If a file is deleted later by applying a new version edit, then calling `GetMinOldestBlobFileNumber()` will not see the obsoleted table file. Therefore, its oldest_blob_file_number will not be counted. It is OK if the blob file is missing.

Reply:
Right, makes sense, thanks

Reply:
Since the per-cf missing_blob_file_number_high will never decrease, once we encounter the first missing blob, we will try to create a version for this column family every time we detect there is no missing file (the missing blob file(s) are below the min_oldest_blob_file_num).